File: packages/rn-sql-connect/android/src/main/kotlin/com/rnsqlconnect/NormalizedError.kt
package com.rnsqlconnect

import com.google.firebase.dataconnect.DataConnectOperationException
import com.google.firebase.dataconnect.DataConnectPathSegment
import org.json.JSONArray
import org.json.JSONObject

/**
 * A platform-neutral error shape.
 *
 * Android throws a tree of exceptions while Apple platforms expose four
 * unrelated structs, so the two taxonomies do not line up. Both sides normalise
 * into this shape and keep the original code in `nativeCode` for debugging.
 */
internal data class NormalizedError(
  val code: String,
  val message: String,
  val operationName: String?,
  val graphQLErrors: List<GraphQLErrorInfo> = emptyList(),
  val partialData: Any? = null,
  val nativeCode: String? = null,
) {
  fun toJson(): JSONObject {
    val json = JSONObject()
    json.put("code", code)
    json.put("message", message)
    json.put("operationName", operationName ?: JSONObject.NULL)
    val errors = JSONArray()
    graphQLErrors.forEach { info ->
      val entry = JSONObject()
      entry.put("message", info.message)
      val path = JSONArray()
      info.path.forEach { segment -> path.put(segment) }
      entry.put("path", path)
      errors.put(entry)
    }
    json.put("graphQLErrors", errors)
    json.put("partialData", JsonBridge.toJsonNode(partialData))
    json.put("nativeCode", nativeCode ?: JSONObject.NULL)
    return json
  }

  companion object {
    /**
     * Maps a throwable coming out of the SDK.
     *
     * Exception types are checked before message text. Text matching is a last
     * resort here because the SDK does not expose the gRPC status: it breaks on
     * any wording change upstream, and it misfires on a message that happens to
     * contain a status name. An unrecognised error becomes `internal` with the
     * original text preserved, which is more useful than guessing.
     */
    fun from(throwable: Throwable, operationName: String?): NormalizedError {
      // A CACHE_ONLY miss arrives as CachedDataNotFoundException. Apple platforms
      // report that situation as `cache-miss`, so without this branch the two
      // platforms disagree about what a cold cache looks like: text matching
      // alone lands it in `internal`, because "cached data not found" does not
      // contain the literal "NOT_FOUND".
      //
      // Matched by class NAME rather than by type on purpose: the class is
      // `internal` in the Firebase SDK, so `is CachedDataNotFoundException` does
      // not compile outside that module. A name is still a far stabler signal
      // than a message, and being internal also means upstream may rename it
      // without calling that a breaking change - hence the fall-through below
      // rather than anything that depends on this matching.
      if (throwable::class.java.simpleName == "CachedDataNotFoundException") {
        return NormalizedError(
          code = "cache-miss",
          message = throwable.message
            ?: "CACHE_ONLY found nothing cached. Note that maxAge defaults to 0, which caches " +
            "responses but never serves them; raise it to read from the cache.",
          operationName = operationName,
          nativeCode = throwable::class.java.simpleName,
        )
      }

      if (throwable is DataConnectOperationException) {
        val response = throwable.response
        val infos = response.errors.map { error ->
          GraphQLErrorInfo(
            message = error.message,
            path = error.path.map { segment ->
              when (segment) {
                is DataConnectPathSegment.Field -> segment.field
                is DataConnectPathSegment.ListIndex -> segment.index
              }
            },
          )
        }
        val hasData = response.data != null || response.rawData != null
        return NormalizedError(
          code = if (hasData) "partial-error" else classify(throwable),
          message = throwable.message ?: "Data Connect operation failed",
          operationName = operationName,
          graphQLErrors = infos,
          partialData = response.data ?: response.rawData,
          nativeCode = throwable::class.java.simpleName,
        )
      }

      return NormalizedError(
        code = classify(throwable),
        message = throwable.message ?: throwable.toString(),
        operationName = operationName,
        nativeCode = throwable::class.java.simpleName,
      )
    }

    private fun classify(throwable: Throwable): String {
      val text = buildString {
        append(throwable.message ?: "")
        var cause = throwable.cause
        var depth = 0
        while (cause != null && depth < 5) {
          append(' ')
          append(cause.message ?: "")
          cause = cause.cause
          depth += 1
        }
      }.uppercase()

      return when {
        text.contains("UNAUTHENTICATED") -> "unauthenticated"
        text.contains("PERMISSION_DENIED") -> "unauthorized"
        text.contains("NOT_FOUND") -> "not-found"
        text.contains("INVALID_ARGUMENT") -> "invalid-argument"
        text.contains("UNAVAILABLE") || text.contains("DEADLINE_EXCEEDED") -> "unavailable"
        text.contains("CANCELLED") -> "cancelled"
        else -> "internal"
      }
    }
  }
}

internal data class GraphQLErrorInfo(
  val message: String,
  val path: List<Any>,
)

File: packages/rn-sql-connect/ios/NormalizedError.swift
import FirebaseDataConnect
import Foundation

/// Platform-neutral error shape.
///
/// Apple platforms expose four unrelated error structs while Android throws a
/// tree of exceptions, so neither taxonomy maps onto the other. Both sides
/// normalise into this shape and keep the platform detail in `nativeCode`.
struct NormalizedError {
  let code: String
  let message: String
  let operationName: String?
  var graphQLErrors: [[String: Any]] = []
  var partialDataJson: String?
  var nativeCode: String?

  private func dictionary() -> [String: Any] {
    var payload: [String: Any] = [
      "code": code,
      "message": message,
      "graphQLErrors": graphQLErrors,
    ]
    payload["operationName"] = operationName ?? NSNull()
    payload["nativeCode"] = nativeCode ?? NSNull()
    return payload
  }

  /// JSON handed to `reject` as `userInfo.details`.
  func detailsJson() -> String {
    var payload = dictionary()
    // `rawJsonData` is already JSON text, so it is spliced in rather than
    // re-encoded, which would turn the object into an escaped string.
    let base = Self.encode(payload)
    guard let partialDataJson else {
      return Self.splice(base, key: "partialData", json: "null")
    }
    payload["partialData"] = nil
    return Self.splice(base, key: "partialData", json: partialDataJson)
  }

  /// Payload pushed to JS for a failing subscription update.
  func eventPayload() -> String {
    "{\"error\":\(detailsJson())}"
  }

  private static func encode(_ payload: [String: Any]) -> String {
    guard
      let data = try? JSONSerialization.data(withJSONObject: payload, options: [.sortedKeys]),
      let text = String(data: data, encoding: .utf8)
    else {
      return "{\"code\":\"internal\",\"message\":\"error encoding failed\"}"
    }
    return text
  }

  private static func splice(_ json: String, key: String, json value: String) -> String {
    guard json.hasSuffix("}") else { return json }
    let body = String(json.dropLast())
    let separator = body.hasSuffix("{") ? "" : ","
    return "\(body)\(separator)\"\(key)\":\(value)}"
  }

  static func from(_ error: Error, operationName: String?) -> NormalizedError {
    let unwrapped: Error
    if let anyError = error as? AnyDataConnectError {
      unwrapped = anyError.dataConnectError
    } else {
      unwrapped = error
    }

    if let operationError = unwrapped as? DataConnectOperationError {
      let response = operationError.response
      let errors: [[String: Any]] = (response?.errors ?? []).map { info in
        [
          "message": info.message,
          "path": (info.path ?? []).map { segment -> Any in
            switch segment {
            case let .field(name): return name
            case let .listIndex(index): return index
            }
          },
        ]
      }
      let hasData = response?.rawJsonData != nil
      return NormalizedError(
        code: hasData ? "partial-error" : classify(operationError.message ?? ""),
        message: operationError.message ?? "Data Connect operation failed",
        operationName: operationName,
        graphQLErrors: errors,
        partialDataJson: response?.rawJsonData,
        nativeCode: "DataConnectOperationError"
      )
    }

    if let initError = unwrapped as? DataConnectInitError {
      return NormalizedError(
        code: "not-configured",
        message: initError.message ?? "Data Connect initialisation failed",
        operationName: operationName,
        nativeCode: "DataConnectInitError.\(initError.code)"
      )
    }

    if let codecError = unwrapped as? DataConnectCodecError {
      return NormalizedError(
        code: "internal",
        message: codecError.message ?? "Data Connect could not decode the response",
        operationName: operationName,
        nativeCode: "DataConnectCodecError.\(codecError.code)"
      )
    }

    if let internalError = unwrapped as? DataConnectInternalError {
      let message = internalError.message ?? "Data Connect internal error"
      return NormalizedError(
        code: classify(message),
        message: message,
        operationName: operationName,
        nativeCode: "DataConnectInternalError.\(internalError.code)"
      )
    }

    let description = String(describing: unwrapped)
    return NormalizedError(
      code: classify(description),
      message: (unwrapped as NSError).localizedDescription,
      operationName: operationName,
      nativeCode: String(describing: type(of: unwrapped))
    )
  }

  /// gRPC status names are matched from the text because the SDK does not
  /// expose the status. Best effort by design: anything unrecognised stays
  /// `internal` with the original message intact.
  private static func classify(_ text: String) -> String {
    let upper = text.uppercased()
    if upper.contains("UNAUTHENTICATED") { return "unauthenticated" }
    if upper.contains("PERMISSION_DENIED") || upper.contains("PERMISSIONDENIED") {
      return "unauthorized"
    }
    if upper.contains("NOT_FOUND") || upper.contains("NOTFOUND") { return "not-found" }
    if upper.contains("INVALID_ARGUMENT") { return "invalid-argument" }
    if upper.contains("UNAVAILABLE") || upper.contains("DEADLINE_EXCEEDED") { return "unavailable" }
    if upper.contains("CANCELLED") || upper.contains("CANCELED") { return "cancelled" }
    return "internal"
  }
}
